matrix_content_scanner/scanner/scanner.py

```diff
@@ -93,12 +150,56 @@ async def scan_file(
             media_content = self._decrypt_file(media_content, metadata)
 
         # Write the file to disk.
-        file_path = self._write_file_to_disk(media_path, media_content)
+        try:
+            file_path = self._write_file_to_disk(media_path, media_content)
+        except FileDirtyError as e:
+            # A FileDirtyError will be raised if the media path is built in a way that
+            # can lead to writing outside of the configured directory.
+            self._result_cache[cache_key] = CacheEntry(
+                result=False,
+                info=e.info,
+            )
+            raise
 
         # Scan the file and see if the result is positive or negative.
         exit_code = self._run_scan(file_path)
         result = exit_code == 0
 
+        # If the exit code isn't part of the ones we should ignore, cache the result.
+        if (
+            self._exit_codes_to_ignore is None
+            or exit_code not in self._exit_codes_to_ignore
+        ):
+            logger.info("Caching result %s", result)
+
+            cached_media = media if result is True else None
+
+            if (
+                self._max_size_to_cache is not None
+                and len(media.content) > self._max_size_to_cache
+            ):
+                # Don't cache the file's content if it exceeds the maximum allowed file
+                # size, to minimise memory usage.
+                logger.info(
+                    "File content has size %s, which is more than %s, not caching content",
+                    format_size(len(media.content)),
+                    format_size(self._max_size_to_cache),
+                )
```

Comment:
seems like we should only do this if we were going to cache the media in the first place? (ie, `cached_media is not None`)